packages/flutter/test/widgets/form_test.dart

```diff
@@ -1457,4 +1457,38 @@ void main() {
 
     expect(focusNode2.hasFocus, isTrue);
   });
+
+  testWidgets('AutovalidateMode.always should validate on first build', (WidgetTester tester) async {
+    String errorText(String? value) => '$value/error';
+
+    await tester.pumpWidget(
+      MaterialApp(
+        theme: ThemeData(),
+        home: Center(
+          child: Form(
+            autovalidateMode: AutovalidateMode.always,
+            child: Material(
+              child: Column(
+                children: <Widget>[
+                  TextFormField(
+                    initialValue: 'foo',
+                    validator: errorText,
+                  ),
+                  TextFormField(
+                    initialValue: 'bar',
+                    validator: errorText,
+                  ),
+                ],
+              ),
+            ),
+          ),
+        )
+      )
+    );
+
+    await tester.pump();
```

Comment:
I'm still not convinced that we can't get this into the correct state in the first frame. Can you see if you can get this test to pass after removing this `pump`?

I would think that when `_errorText` is assigned in `initState`, you could set its initial value to something like this:

```dart
widget.forceErrorText ?? widget.validator?.call(_value)
```

Would that work?

Reply:
I believe the post-frame callback is the non-breaking way to do it, since the `validator` could potentially contain a `setState` call of the parent widget.

example: https://dartpad.dev/?id=62b1d1bce1cd79fa041135aab38d6d93

Reply:
> I believe the post-frame callback is the non-breaking way to do it, since the `validator` could potentially contain a `setState` call of the parent widget.
> 
> example: https://dartpad.dev/?id=62b1d1bce1cd79fa041135aab38d6d93

Exactly. The post-frame callback breaks this logic, because by removing it and running only this test, the `pump` becomes unnecessary.

Reply:
Did you try it and it caused failures? The validator function sounds like something that shouldn't have side effects but maybe it's too late to make any changes like this if it's going to break everyone.

I tried removing the postFrameCallback but it still crashes if you call setState in `validator`.

<details>

<summary>Diff</summary>

```diff
diff --git a/packages/flutter/lib/src/widgets/form.dart b/packages/flutter/lib/src/widgets/form.dart
index 98c63fe884..0b1f90c1bd 100644
--- a/packages/flutter/lib/src/widgets/form.dart
+++ b/packages/flutter/lib/src/widgets/form.dart
@@ -550,7 +550,7 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
   late T? _value = widget.initialValue;
   // Marking it as late, so it can be registered
   // with the value provided by [forceErrorText].
-  late final RestorableStringN _errorText;
+  RestorableStringN? _errorText;
   final RestorableBool _hasInteractedByUser = RestorableBool(false);
   final FocusNode _focusNode = FocusNode();
 
@@ -564,10 +564,10 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
   /// This property is automatically updated when [validate] is called and the
   /// [FormField.validator] callback is invoked, or If [FormField.forceErrorText] is set
   /// directly to a non-null value.
-  String? get errorText => _errorText.value;
+  String? get errorText => _errorText!.value;
 
   /// True if this field has any validation errors.
-  bool get hasError => _errorText.value != null;
+  bool get hasError => _errorText!.value != null;
 
   /// Returns true if the user has modified the value of this field.
   ///
@@ -597,7 +597,7 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
     setState(() {
       _value = widget.initialValue;
       _hasInteractedByUser.value = false;
-      _errorText.value = null;
+      _errorText!.value = null;
     });
     Form.maybeOf(context)?._fieldDidChange();
   }
@@ -619,14 +619,14 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
 
   void _validate() {
     if (widget.forceErrorText != null) {
-      _errorText.value = widget.forceErrorText;
+      _errorText!.value = widget.forceErrorText;
       // Skip validating if error is forced.
       return;
     }
     if (widget.validator != null) {
-      _errorText.value = widget.validator!(_value);
+      _errorText!.value = widget.validator!(_value);
     } else {
-      _errorText.value = null;
+      _errorText!.value = null;
     }
   }
 
@@ -662,7 +662,7 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
 
   @override
   void restoreState(RestorationBucket? oldBucket, bool initialRestore) {
-    registerForRestoration(_errorText, 'error_text');
+    registerForRestoration(_errorText!, 'error_text');
     registerForRestoration(_hasInteractedByUser, 'has_interacted_by_user');
   }
 
@@ -672,23 +672,28 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
     super.deactivate();
   }
 
-  @override
-  void initState() {
-    super.initState();
-    _errorText = RestorableStringN(widget.forceErrorText);
-  }
-
   @override
   void didUpdateWidget(FormField<T> oldWidget) {
     super.didUpdateWidget(oldWidget);
     if (widget.forceErrorText != oldWidget.forceErrorText) {
-      _errorText.value = widget.forceErrorText;
+      _errorText!.value = widget.forceErrorText;
     }
   }
 
   @override
   void didChangeDependencies() {
+    final AutovalidateMode? autovalidateMode = Form.maybeOf(context)?.widget.autovalidateMode;
+    if (_errorText == null) {
+      if (autovalidateMode == AutovalidateMode.always) {
+        final String? errorText = widget.forceErrorText ?? widget.validator?.call(_value);
+        _errorText = RestorableStringN(errorText);
+      } else {
+        _errorText = RestorableStringN(widget.forceErrorText);
+      }
+    }
+
     super.didChangeDependencies();
+
     switch (Form.maybeOf(context)?.widget.autovalidateMode) {
       case AutovalidateMode.always:
         WidgetsBinding.instance.addPostFrameCallback((_) {
@@ -707,7 +712,7 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
 
   @override
   void dispose() {
-    _errorText.dispose();
+    _errorText!.dispose();
     _focusNode.dispose();
     _hasInteractedByUser.dispose();
     super.dispose();
diff --git a/packages/flutter/test/widgets/form_test.dart b/packages/flutter/test/widgets/form_test.dart
index 13c9ba3263..466506ebe5 100644
--- a/packages/flutter/test/widgets/form_test.dart
+++ b/packages/flutter/test/widgets/form_test.dart
@@ -1486,7 +1486,7 @@ void main() {
       )
     );
 
-    await tester.pump();
+    //await tester.pump();
 
     expect(find.text(errorText('foo')), findsOneWidget);
     expect(find.text(errorText('bar')), findsOneWidget);

```

</details>

Reply:
Sorry for the delay. I tried a few approaches as well, but they didn't work.😔

Reply:
Sounds like this is our best option then, thanks for checking!

Nit: Maybe add `expect`s above this pump to check that the error text does not yet appear, and explain?

```dart
    // The validation happens in a post frame callback, so the error
    // doesn't show up until the second frame.
    expect(find.text(errorText('foo')), findsNothing);
    expect(find.text(errorText('bar')), findsNothing);

    await tester.pump();

   ...
```